pkg/model/types.go

```diff
@@ -41,61 +39,127 @@ type Service struct {
+	ServicePort  Port
+	Attributes   map[string]string
+}
+
+type Port struct {
+	Name       string
 	Port       int32
-	Attributes map[string]string
+	TargetPort string
+	Protocol   string // TCP, UDP, SCTP
 }
 
+// Cloudmap Instances IP and Port is supposed to be AWS_INSTANCE_IPV4 and AWS_INSTANCE_PORT
+// Rest are custom attributes
 const (
-	Ipv4Attr = "AWS_INSTANCE_IPV4"
-	PortAttr = "AWS_INSTANCE_PORT"
+	EndpointIpv4Attr      = "AWS_INSTANCE_IPV4"
+	EndpointPortAttr      = "AWS_INSTANCE_PORT"
+	EndpointPortNameAttr  = "ENDPOINT_PORT_NAME"
+	EndpointProtocolAttr  = "ENDPOINT_PROTOCOL"
+	ServicePortNameAttr   = "SERVICE_PORT_NAME"
+	ServicePortAttr       = "SERVICE_PORT"
+	ServiceTargetPortAttr = "SERVICE_TARGET_PORT"
+	ServiceProtocolAttr   = "SERVICE_PROTOCOL"
 )
 
 // NewEndpointFromInstance converts a Cloud Map HttpInstanceSummary to an endpoint.
-func NewEndpointFromInstance(inst *types.HttpInstanceSummary) (*Endpoint, error) {
+func NewEndpointFromInstance(inst *types.HttpInstanceSummary) (endpointPtr *Endpoint, err error) {
 	endpoint := Endpoint{
 		Id:         *inst.InstanceId,
-		Attributes: make(map[string]string, 0),
+		Attributes: make(map[string]string),
+	}
+	attributes := make(map[string]string)
+	for key, value := range inst.Attributes {
+		attributes[key] = value
 	}
 
-	if ipv4, hasIp := inst.Attributes[Ipv4Attr]; hasIp {
-		endpoint.IP = ipv4
-	} else {
-		return nil, errors.New(fmt.Sprintf("cannot convert service instance %s to endpoint without IP address",
-			*inst.InstanceId))
+	// Remove and set the IP, Port, Port
+	if endpoint.IP, err = removeStringAttr(attributes, EndpointIpv4Attr); err != nil {
+		return nil, err
 	}
 
-	if portStr, hasPort := inst.Attributes[PortAttr]; hasPort {
-		port, parseError := strconv.ParseUint(portStr, 10, 16)
+	if endpoint.EndpointPort, err = endpointPortFromAttr(attributes); err != nil {
+		return nil, err
+	}
 
-		if parseError != nil {
-			return nil, parseError
-		}
+	if endpoint.ServicePort, err = servicePortFromAttr(attributes); err != nil {
+		return nil, err
+	}
 
-		endpoint.Port = int32(port)
-	} else {
-		return nil, errors.New(fmt.Sprintf("cannot convert service instance %s to endpoint without port",
-			*inst.InstanceId))
+	// Add the remaining attributes
+	endpoint.Attributes = attributes
+
+	return &endpoint, err
+}
+
+func endpointPortFromAttr(attributes map[string]string) (port Port, err error) {
+	port = Port{}
+	if port.Name, err = removeStringAttr(attributes, EndpointPortNameAttr); err != nil {
+		return port, err
 	}
+	if port.Port, err = removeIntAttr(attributes, EndpointPortAttr); err != nil {
+		return port, err
+	}
+	if port.Protocol, err = removeStringAttr(attributes, EndpointProtocolAttr); err != nil {
+		return port, err
+	}
+	return port, err
+}
 
-	for key, val := range inst.Attributes {
-		if key != Ipv4Attr && key != PortAttr {
-			endpoint.Attributes[key] = val
-		}
+func servicePortFromAttr(attributes map[string]string) (port Port, err error) {
```

Comment:
Re: multiple clusters + mixed schemas, since we don't support multiple clusters exporting a single service yet ([upcoming 0.3 XL story](https://github.com/aws/aws-cloud-map-mcs-controller-for-k8s/issues/47)), we don't have to worry about schema collisions within Cloud Map, as long as an upgrade cleans up the old instances. We still have to think about an old import controller handling an upgraded export controller. I believe we're okay in this regard with this change?

If all else fails for the sake of velocity we could suggest running the janitor for 0.1 -> 0.2 upgrade, although I don't think it's so dire.